### PR TITLE
heartbeat challenge

### PR DESCRIPTION
**Pull request by johannbarbie**
resolves https://github.com/leapdao/leap-contracts/issues/257

### CODE REVIEW

**Comment by troggy** [comment]
good todos, worth extracting as separate issues

File: contracts/PoaOperator.sol [hunk 1]
```diff
@@ -291,6 +304,126 @@ contract PoaOperator is Adminable {
...
+    beatChallenges[slot.signer] = BeatChallenge({
+      challenger: msg.sender,
+      openTime: now,
+      openPeriodHash: tip
+    });
+  }
+
+  // In case of an invalid challenge, can submit a proof of heartbeat
+  // TODO: Is the signer we challenged still in control of the Slot? Depends on timeout-time (currently casChallangeDuration) at the least.
+  function respondBeat(
```
Comment (johannbarbie) [inline contracts/PoaOperator.sol]: this needs front-running protection. either include address in mgs.data, or send funds to tx signer?

File: contracts/PoaOperator.sol [hunk 2]
```diff
@@ -291,6 +304,126 @@ contract PoaOperator is Adminable {
...
+    bytes32[] memory _inclusionProof,
+    bytes32[] memory _walkProof,
```
Comment (troggy) [inline contracts/PoaOperator.sol]: @TheReturnOfJan could you please more comments about what is `walkProof` ?

File: contracts/PoaOperator.sol (hunk 1 above)
Comment (TheReturnOfJan) [reply contracts/PoaOperator.sol]: Fixed in new commit.

File: contracts/PoaOperator.sol (hunk 2 above)
Comment (TheReturnOfJan) [reply contracts/PoaOperator.sol]: @troggy new commit should be a bit clearer.

**Pull request merged**